src/Datadog.Trace/Containers/ContainerInfo.cs

```diff
@@ -0,0 +1,98 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text.RegularExpressions;
+
+namespace Datadog.Trace.Containers
+{
+    /// <summary>
+    /// Utility class with methods to interact with container hosts.
+    /// </summary>
+    internal static class ContainerInfo
+    {
+        private const string ControlGroupsFilePath = "/proc/self/cgroup";
+
+        private const string LineRegex = @"^(?:\d+):(?:[^:]*):(.+)$";
+
+        private const string ContainerIdRegex = @"(?:pod)?" +
+                                                @"([0-9a-f]{8}[-_][0-9a-f]{4}[-_][0-9a-f]{4}[-_][0-9a-f]{4}[-_][0-9a-f]{12}|[0-9a-f]{64})" +
+                                                @"(?:\.scope|\.slice)?$";
+
+        /// <summary>
+        /// Gets the id of the container executing the code.
+        /// Return <c>null</c> if code is not executing inside a supported container.
+        /// </summary>
+        /// <returns>The container id or <c>null</c>.</returns>
+        public static string GetContainerId()
```

Comment:
Would it be worth caching the container id in case we ever call for it several times?

Reply:
Added a `Lazy<T>` to cache this. Thanks!